File: key/keyext.interactionlog/src/main/java/org/key_project/ui/interactionlog/model/Interaction.java
package org.key_project.ui.interactionlog.model;

import org.key_project.ui.interactionlog.algo.InteractionVisitor;

import javax.swing.*;
import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlTransient;
import java.awt.*;
import java.io.Serializable;
import java.util.Date;

/**
 * @author weigl
 */

@XmlTransient
@XmlAccessorType(XmlAccessType.FIELD)
public abstract class Interaction implements Serializable {
    private static final long serialVersionUID = -4261703397513027869L;

    @XmlTransient
    protected InteractionGraphicStyle graphicalStyle = new InteractionGraphicStyle();

    @XmlAttribute
    private Date created = new Date();

    @XmlAttribute
    private boolean favoured = false;

    public Date getCreated() {
        return created;
    }

    public void setCreated(Date created) {
        this.created = created;
    }

    public boolean isFavoured() {
        return favoured;
    }

    public void setFavoured(boolean favoured) {
        this.favoured = favoured;
    }

    public InteractionGraphicStyle getGraphicalStyle() {
        return graphicalStyle;
    }

    public abstract <T> T accept(InteractionVisitor<T> visitor);

    public static class InteractionGraphicStyle {
        private Icon icon;
        private Color backgroundColor, foregroundColor;

        public Icon getIcon() {
            return icon;
        }

        public void setIcon(Icon icon) {
            this.icon = icon;
        }

        public Color getBackgroundColor() {
            return backgroundColor;
        }

        public void setBackgroundColor(Color backgroundColor) {
            this.backgroundColor = backgroundColor;
        }

        public Color getForegroundColor() {
            return foregroundColor;
        }

        public void setForegroundColor(Color foregroundColor) {
            this.foregroundColor = foregroundColor;
        }
    }
}

File: key/keyext.interactionlog/src/main/kotlin/org/key_project/ui/interactionlog/api/Interaction.kt
package org.key_project.ui.interactionlog.api

import org.key_project.ui.interactionlog.api.Markdownable
import org.key_project.ui.interactionlog.api.Reapplicable
import org.key_project.ui.interactionlog.api.Scriptable

import javax.swing.*
import javax.xml.bind.annotation.XmlAccessType
import javax.xml.bind.annotation.XmlAccessorType
import javax.xml.bind.annotation.XmlAttribute
import javax.xml.bind.annotation.XmlTransient
import java.awt.*
import java.io.Serializable
import java.util.Date

/**
 * @author weigl
 */
@XmlTransient
@XmlAccessorType(XmlAccessType.FIELD)
abstract class Interaction : Serializable, Markdownable, Scriptable, Reapplicable {
    @XmlTransient
    var graphicalStyle = InteractionGraphicStyle()
        protected set

    @XmlAttribute
    var created = Date()

    @XmlAttribute
    var isFavoured = false

    class InteractionGraphicStyle {
        var icon: Icon? = null
        var backgroundColor: Color? = null
        var foregroundColor: Color? = null
    }
}
